R/pkg/R/DataFrame.R

```diff
@@ -2200,4 +2200,107 @@ setMethod("coltypes",
             rTypes[naIndices] <- types[naIndices]
 
             rTypes
-          })
\ No newline at end of file
+          })
+
+#' Display the structure of a DataFrame, including column names, column types, as well as a
+#' a small sample of rows.
+#' @name str
+#' @title Compactly display the structure of a dataset
+#' @rdname str_data_frame
+#' @family dataframe_funcs
+#' @param x a DataFrame
+#' @examples \dontrun{
+#'
+#' # Create a DataFrame from the Iris dataset
+#' irisDF <- createDataFrame(sqlContext, iris)
+#' 
+#' # Show the structure of the DataFrame
+#' str(irisDF)
+#' 
+#' }
+setMethod("str", signature="DataFrame", definition=
+            function(object) {
+
+              # A synonym for easily concatenating strings
+              "%++%" <- function(x, y) {
+                paste(x, y, sep = "")
+              }
+
+              # TODO: These could be made global parameters, though in R it's not the case
+              DEFAULT_HEAD_ROWS <- 6
+              MAX_CHAR_PER_ROW <- 120
+              MAX_COLS <- 100
+
+              # Get the column names and types of the DataFrame
+              names <- names(object)
+              types <- coltypes(object)
+
+              # Get the number of rows.
+              # TODO: Ideally, this should be cached
+              cachedCount <- nrow(object)
+
+              # Get the first elements of the dataset. Limit number of columns accordingly
+              dataFrame <- if (ncol(object) > MAX_COLS) {
+                             head(object[, c(1:MAX_COLS)], DEFAULT_HEAD_ROWS)
+                           } else {
+                             head(object, DEFAULT_HEAD_ROWS)
+                           }
+
+              # The number of observations will be displayed only if the number
+              # of rows of the dataset has already been cached.
+              if (!is.null(cachedCount)) {
+                cat("'" %++% class(object) %++% "': " %++% cachedCount %++% " obs. of " %++%
+                      length(names) %++% " variables:\n")
+              } else {
+                cat("'" %++% class(object) %++% "': " %++% length(names) %++% " variables:\n")
+              }
+
+              # Whether the ... should be printed at the end of each row
+              ellipsis <- FALSE
+
+              # Add ellipsis (i.e., "...") if there are more rows than shown
+              if (!is.null(cachedCount)) {
+                if (nrow(object) > DEFAULT_HEAD_ROWS) {
+                  ellipsis <- TRUE
+                }
+              }
+
+              if (nrow(dataFrame) > 0) {
+                for (i in 1 : ncol(dataFrame)) {
+                  firstElements <- ""
+
+                  # Get the first elements for each column
+                  if (types[i] == "chr") {
```

Comment:
I get the types from coltypes(). Now to get the short types, I added an environment called SHORT_TYPES. In there, "character" is mapped to "chr", "logical" to "logi" and so forth.